File: .claude/hooks/session-start-stash-warning.mjs
#!/usr/bin/env node
// @hookstack session-start-stash-warning
// SessionStart: avertit si des stashs Git non récupérés traînent depuis plusieurs jours
import { execSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const DAYS = 3;

function defaultExec(cmd) {
  try { return execSync(cmd, { encoding: 'utf8', timeout: 5_000 }).trim(); } catch { return ''; }
}

export function run({ exec = defaultExec, now = () => Date.now() } = {}) {
  const stashList = exec('git stash list --format="%gd|%ci|%gs"');
  if (!stashList) return null;

  const currentTime = now();
  const threshold = DAYS * 24 * 60 * 60 * 1000;

  const stale = stashList
    .split('\n')
    .map((line) => {
      const [ref, date, ...msgParts] = line.split('|');
      const age = currentTime - new Date(date).getTime();
      return { ref, date: date?.trim(), msg: msgParts.join('|').trim(), age };
    })
    .filter((s) => s.age > threshold);

  if (stale.length === 0) return null;

  const lines = [
    `## ⚠️  Stashs Git oubliés (${stale.length})`,
    `- ${stale.length} stash(s) de plus de ${DAYS} jours détecté(s) :`,
  ];

  for (const s of stale.slice(0, 5)) {
    const days = Math.floor(s.age / (24 * 60 * 60 * 1000));
    lines.push(`  - \`${s.ref}\` (${days}j) — ${s.msg}`);
  }

  lines.push(`- Utilisez \`git stash list\` pour les voir, \`git stash pop\` ou \`git stash drop\` pour les gérer.`);

  return `${lines.join('\n')}\n`;
}

/* v8 ignore next 4 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const result = run();
  if (result) process.stdout.write(result);
}

File: .claude/hooks/session-start-stash-warning.py
#!/usr/bin/env python3
# @hookstack session-start-stash-warning
"""Warns about stale git stashes older than 3 days (SessionStart)."""
import subprocess
import sys
import time
from datetime import datetime, timezone

DAYS = 3


def _exec(cmd):
    try:
        return subprocess.run(
            cmd, shell=True, capture_output=True, text=True, timeout=5, check=True
        ).stdout.strip()
    except Exception:
        return ""


def _now_ms():
    return int(time.time() * 1000)


def _parse_iso(value):
    try:
        dt = datetime.fromisoformat(value.strip())
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt.timestamp() * 1000
    except Exception:
        return None


def run(*, exec_cmd=None, now=None):
    if exec_cmd is None:
        exec_cmd = _exec
    if now is None:
        now = _now_ms

    stash_list = exec_cmd('git stash list --format="%gd|%ci|%gs"')
    if not stash_list:
        return None

    current = now()
    threshold = DAYS * 24 * 60 * 60 * 1000

    stale = []
    for line in stash_list.split("\n"):
        parts = line.split("|")
        if len(parts) < 2:
            continue
        ref = parts[0]
        date = parts[1].strip()
        msg = "|".join(parts[2:]).strip()
        ts = _parse_iso(date)
        if ts is None:
            continue
        age = current - ts
        if age > threshold:
            stale.append({"ref": ref, "msg": msg, "age": age})

    if not stale:
        return None

    lines = [
        f"## ⚠️  Stashs Git oubliés ({len(stale)})",
        f"- {len(stale)} stash(s) de plus de {DAYS} jours détecté(s) :",
    ]

    for s in stale[:5]:
        days = int(s["age"] // (24 * 60 * 60 * 1000))
        lines.append(f"  - `{s['ref']}` ({days}j) — {s['msg']}")

    lines.append(
        "- Utilisez `git stash list` pour les voir, `git stash pop` ou `git stash drop` pour les gérer."
    )

    return "\n".join(lines) + "\n"


if __name__ == "__main__":
    result = run()
    if result:
        sys.stdout.write(result)
